tests/serac_odes.cpp

```diff
@@ -0,0 +1,645 @@
+
+IterativeSolverOptions linear_options{.rel_tol     = 1.0e-12,
+                                      .abs_tol     = 1.0e-12,
+                                      .print_level = -1,
+                                      .max_iter    = 10,
+                                      .lin_solver  = LinearSolver::CG,
+                                      .prec        = {}};
+
+NonlinearSolverOptions nonlinear_options{.rel_tol = 1.0e-12, .abs_tol = 1.0e-12, .max_iter = 10, .print_level = -1};
+
+mfem::DenseMatrix M = []() {
+  mfem::DenseMatrix M(3, 3);
+  M(0, 0) = 2.0;
+  M(0, 1) = 1.0;
+  M(0, 2) = 0.0;
+
+  M(1, 0) = 1.0;
+  M(1, 1) = 4.0;
+  M(1, 2) = 1.0;
+
+  M(2, 0) = 0.0;
+  M(2, 1) = 1.0;
+  M(2, 2) = 2.0;
+  return M;
+}();
+
+mfem::DenseMatrix C = []() {
+  mfem::DenseMatrix C(3, 3);
+  C(0, 0) = 1.0;
+  C(0, 1) = 0.0;
+  C(0, 2) = 0.0;
+
+  C(1, 0) = 0.0;
+  C(1, 1) = 1.0;
+  C(1, 2) = 0.0;
+
+  C(2, 0) = 0.0;
+  C(2, 1) = 0.0;
+  C(2, 2) = 1.0;
+  return C;
+}();
+
+std::function stiffness_linear = [](const mfem::Vector & /*x*/) -> mfem::DenseMatrix {
+  mfem::DenseMatrix K(3, 3);
+  K(0, 0) = 1.0;
+  K(0, 1) = -1.0;
+  K(0, 2) = 0.0;
+
+  K(1, 0) = -1.0;
+  K(1, 1) = 2.0;
+  K(1, 2) = -1.0;
+
+  K(2, 0) = 0.0;
+  K(2, 1) = -1.0;
+  K(2, 2) = 1.0;
+  return K;
+};
+
+std::function internal_force_linear = [](const mfem::Vector& x) -> mfem::Vector { return stiffness_linear(x) * x; };
+
+std::function stiffness_nonlinear = [](const mfem::Vector& x) -> mfem::DenseMatrix {
+  mfem::DenseMatrix K(3, 3);
+  K(0, 0) = 1.0;
+  K(0, 1) = -1.0;
+  K(0, 2) = 0.0;
+
+  K(1, 0) = -1.0;
+  K(1, 1) = 2.0 - 2 * x(1) + x(2);
+  K(1, 2) = -1.0 + x(1);
+
+  K(2, 0) = 0.0;
+  K(2, 1) = -1.0 + x(2);
+  K(2, 2) = 1.0 + x(1) + x(2);
+  return K;
+};
+
+std::function internal_force_nonlinear = [](const mfem::Vector& x) -> mfem::Vector {
+  mfem::Vector f(3);
+  f(0) = x(0) - x(1);
+  f(1) = (2.0 - x(1)) * x(1) + (x(1) - 1.0) * x(2) - x(0);
+  f(2) = x(1) * (x(2) - 1.0) + (1.0 + 0.5 * x(2)) * x(2);
+  return f;
+};
+
+mfem::Vector f_ext = []() {
+  mfem::Vector f(3);
+  f(0) = 1.0;
+  f(1) = 2.0;
+  f(2) = 3.0;
+  return f;
+}();
+
+// continuous constraint with continuous derivative
+const auto sine_wave = [](const mfem::Vector& /*x*/, double t) { return 1.0 + sin(2.0 * M_PI * t); };
+
+// continuous constraint with discontinuous derivative
+const auto triangle_wave = [](const mfem::Vector& /*x*/, double t) {
+  return 1.0 + (2.0 / M_PI) * asin(sin(2.0 * M_PI * t));
+};
+
+enum ode_type
+{
+  LINEAR,
+  NONLINEAR
+};
+
+std::string to_string(ode_type o)
+{
+  if (o == LINEAR) return std::string("linear");
+  if (o == NONLINEAR) return std::string("nonlinear");
+  return "unknown";
+}
+
+enum constraint_type
+{
+  UNCONSTRAINED,
+  SINE_WAVE,
+  TRIANGLE_WAVE
+};
+
+std::string to_string(constraint_type c)
+{
+  if (c == UNCONSTRAINED) return std::string("unconstrained");
```

Comment:
Out of curiosity, did you run into something weird when you omitted the explicit constructor call?

Reply:
I did not, I will removing the explicit `std::string(****)` constructors